core/common/src/main/java/alluxio/collections/NonUniqueFieldIndex.java

```diff
@@ -98,4 +111,80 @@ public T getFirst(Object value) {
     Set<T> all = mIndexMap.get(value);
     return all == null ? null : Iterables.getFirst(all, null);
   }
+
+  @Override
+  public Iterator<T> iterator() {
+    return new NonUniqueFieldIndexIterator();
+  }
+
+  @Override
+  public int size() {
+    int totalSize = 0;
+    for (ConcurrentHashSet<T> innerSet : mIndexMap.values()) {
+      totalSize += innerSet.size();
+    }
+    return totalSize;
+  }
+
+  /**
+   * Specialized iterator for {@link NonUniqueFieldIndex}.
+   *
+   * This is needed to support consistent removal from the set and the indices.
+   */
+  private class NonUniqueFieldIndexIterator implements Iterator<T> {
+    /**
+     * Iterator of {@link NonUniqueFieldIndex#mIndexMap}. This Iterator keeps track of the
+     * inner set which is under iteration.
+     */
+    private final Iterator<ConcurrentHashSet<T>> mIndexMapIterator;
+    /**
+     * Iterator inside each inner set. This iterator keeps track of the objects.
+     */
+    private Iterator<T> mIndexIterator;
+    /**
+     * Keeps track of current object. It is used to do the remove.
+     */
+    private T mObject;
+
+    public NonUniqueFieldIndexIterator() {
+      mIndexMapIterator = mIndexMap.values().iterator();
+      mIndexIterator = null;
+      mObject = null;
+    }
+
+    @Override
+    public boolean hasNext() {
```

Comment:
I think the functionality of `hasNext` and `next` have to be modified. https://docs.oracle.com/javase/7/docs/api/java/util/Iterator.html

I think technically, you don't have to call `hasNext` before `next`. Also, I think you can call `hasNext` multiple times before the `next` call. Basically, `hasNext` should be idempotent. Therefore, `next` should return the next value, as well as update state, and `hasNext` should simply check if the next exists.